service/raft/raft_group0.cc

```diff
@@ -424,11 +486,65 @@ future<> raft_group0::leave_group0() {
         co_return;
     }
 
-    if (!joined_group0()) {
-        // TODO: unimplemented upgrade case.
+    auto upgrade_state = (co_await _client.get_group0_upgrade_state()).second;
+    if (upgrade_state == group0_upgrade_state::recovery) {
+        group0_log.warn("leave_group0: in Raft RECOVERY mode, skipping.");
         co_return;
     }
 
+    if (!_feat.supports_raft_cluster_mgmt) {
+        // The Raft feature is not yet enabled.
+        //
+        // In that case we assume the cluster is in a partially-upgraded state (some nodes still use a version
+        // without the raft/group 0 code enabled) and skip the leaving group 0 step.
+        //
+        // In theory we could be wrong: the cluster may have just upgraded but the user managed to call decommission
+        // before we noticed it. Us leaving at this point could cause `upgrade_to_group0`
+        // on other nodes to get stuck because it requires contacting all peers which may include us.
+        //
+        // For this to happen, the following conditions would have to be true:
+        // 1. decommission is called immediately after upgrade, before we notice that the Raft feature is enabled
+        // 2. we don't notice it during decommission - including the long streaming phase - even until we reach
+        //    the `leave_group0` step which is almost at the end of the procedure
+        // 3. some node did notice that the feature is enabled and started `upgrade_to_group0`, using us
+        //    as one of the peers for running group 0 discovery, before we entered LEFT state
+        //
+        // These conditions together give a very unlikely scenario. If it does happen the user can perform
+        // the manual recovery procedure for group 0.
```

Comment:
> Supporting the feature is equivalent to locally entering upgrade_state = synchronize. group0 operations are blocked until use_new_raft_procedures feature is supported locally

`synchronize` means that schema changes are disabled.
That would basically mean a node blocks all schema changes as soon as it starts supporting the feature.
So for example, in a partially upgraded cluster, the nodes that are upgraded don't accept schema changes. I didn't want that.

I also wanted to ensure that A. group 0 can be created and joined and schema can be synchronized before disabling schema changes (there's the dry-run step of synchronizing schema before entering `synchronize` state) and B. we do as much work as we can, such as joining group 0, before we enter `synchronize`, so schema changes are blocked for the smallest amount of time possible.

Reply:
We can introduce an intermediate feature for `synchronize` to achieve small lock window (re: B). It will be supported only after whole cluster is upgraded, or any additional upgrade work which is needed:

```
supports_raft_cluster_mgmt_feature.when_enabled([] {
   co_await drain_local_schema_changes_and_lock();
   synchronize_feature.support();
});
```

Re: A, what's the reason for explicit pull before blocking new schema changes? It seems it makes the whole thing slower, since we need more pulls. We need a pull after synchronize anyway.

Reply:
> We can introduce an intermediate feature for synchronize to achieve small lock window (re: B). It will be supported only after whole cluster is upgraded, or any additional upgrade work which is needed:

We could replace the single `group0_upgrade_state` enum class with two new cluster features. We would also need to implement `feature::support`. Also I don't know how supporting features works with persistence - we need to make sure that a feature can't be "un-supported",  hope we don't need to have a separate persistent state for this anyway.

Then there's the question of how we do recovery, which is currently done through setting `group0_upgrade_state` to a special `recovery` state, from which we can then enter `use_old_procedures` and perform upgrade again. Third cluster feature to ensure we're not in `recovery`? Then this means we basically need to introduce "un-supporting" features.

Doesn't look like cluster features are the right tool for the job to me.

> Re: A, what's the reason for explicit pull before blocking new schema changes? It seems it makes the whole thing slower, since we need more pulls.

We first pull versions, we only pull schema if versions are mismatching. So we don't do wasteful schema pulls if needed.

Now, the point of this step is to improve liveness. Suppose schema is in some weird state that prevents it from synchronizing. We will then block on this step, preventing us from entering `synchronize`, which is good because `synchronize` disables schema changes. If we didn't do this step, we would enter `synchronize` and get stuck inside.
Of course it may happen that we enter this weird state in the short time period between the dry run and entering `synchronize`, but we would have to be incredibly unlucky.

Even if schema is not in some weird state, the dry run makes the lock window shorter, because if we manage to ensure schema is synced now, during the lock window we will only have to verify that versions are the same (unless we're incredibly unlucky).

Reply:
If schema is in a weird state which prevents it from synchronizing, then schema changes will fail too. 

`sync_schema()` should also be a no-op if the schema is synchronized, which should be the common case during upgrade. Adding code to make the window a bit shorter in corner cases is a bad trade-off. 

I still don't see good enough reason for not doing a single `migration_manager::sync_schema()` after entering `synchronize` state globally.

Reply:
> If schema is in a weird state which prevents it from synchronizing, then schema changes will fail too.

I was fixing bugs in customer clusters by modifying schema tables manually on every node and then performing a schema change to force resynchronization / digest recalculation. It wouldn't be good if schema changes were unavailable to me at this point.

> sync_schema() should also be a no-op if the schema is synchronized, which should be the common case during upgrade.

You must check if schema is synchronized - by pulling schema versions from remotes. So it cannot be a no-op.

Reply:
> I still don't see good enough reason for not doing a single migration_manager::sync_schema() after entering synchronize state globally.

https://github.com/scylladb/scylladb/pull/10835#discussion_r943633165

Reply:
> Adding code to make the window a bit shorter in corner cases is a bad trade-off.

As I said - it's not just making the window a bit shorter. It's also ensuring that schema is synchronizable at all before we lock the cluster. It makes the procedure robust.

Reply:
> You must check if schema is synchronized - by pulling schema versions from remotes. So it cannot be a no-op.

True.

> I was fixing bugs in customer clusters by modifying schema tables manually on every node and then performing a schema change to force resynchronization / digest recalculation. It wouldn't be good if schema changes were unavailable to me at this point.

What if this happens during the `synchronize` phase? We need a solution for that too.

Reply:
> As I said - it's not just making the window a bit shorter. It's also ensuring that schema is synchronizable at all before we lock the cluster. It makes the procedure robust.

Yes, good point. Why not invoke the existing `wait_for_schema_agreement()` though, like we do during bootstrap?

Reply:
> What if this happens during the synchronize phase? We need a solution for that too.

We do have a solution for that - namely, the `recovery` state, which one can enter by setting `group0_upgrade_state` in `system.local` to the string `recovery`, then restart Scylla. In this state, schema changes are allowed (they use the old procedures).

You could now argue that since we do have this, it covers all bases - so the synchronization step before entering `synchronize` is unnecessary, since even if it gets stuck, we can enter `recovery`.

And I agree that strictly speaking, it's not necessary. Correctness is preserved without it, and given the fact that the administrator is able to recover the cluster manually, we have liveness too (through admin's manual intervention).

Still, I wanted to make the procedure as robust as possible. I want to cover as many bases as we can without having to enter `recovery` state, because doing that requires extra care.

Besides, it's not like the code becomes much bigger; it's just an extra call to `synchronize_schema` before entering the `synchronize` state inside `do_upgrade_to_group0`.

It's true that on the happy path, this will maybe increase the length of the procedure because of the extra RPC call to fetch the remote schema? But:
- this procedure is executed only once per node (hopefully, assuming we don't need recovery),
- how long does it take to RPC everyone? 300ms? (if we're distributed across the globe) It won't really make a noticeable difference.

Reply:
> Yes, good point. Why not invoke the existing wait_for_schema_agreement() though, like we do during bootstrap?

Unfortunately, `wait_for_schema_agreement` uses the current local state - gossiper's application state - to check whether "schema is in agreement".

But it could happen that a remote node made a schema change, which we didn't learn about yet, and we think that schema is in agreement, but it's "really" not. There could be, for example, a problem with gossiper, causing schema changes to be propagated really slowly.

Unless you mean that this is before `synchronize` state - so we can be non-exact at this point?
Then yes, I guess we could use `wait_for_schema_agreement` before we enter `synchronize` (but we still need the stronger `synchronize_schema` after we enter `synchronize`). However, it wouldn't change the code much - just replace one function call with another (`synchronize_schema` has to stay for the second call) - it would just become a bit less robust.

Reply:
yes, I meant before synchronize state. If we still need the synchronize_schema() mechanism, then it doesn't matter. But there was a point made against my suggestion to get rid of synchronize_schema() that we still need it for his step before the synchronize step. So I'm pointing out that we can call wait_for_schema_agreement() if we do the alternative approach.